File: minimumTimeToBrewPotions.go
func minTime(skill []int, mana []int) int64 {
    n := len(skill)
    f := make([]int64, n)

    for _, x := range mana {
        now := f[0]
        for i := 1; i < n; i++ {
            candidate := now + int64(skill[i-1])*int64(x)
            if candidate > f[i] {
                now = candidate
            } else {
                now = f[i]
            }
        }
        f[n-1] = now + int64(skill[n-1])*int64(x)
        for i := n - 2; i >= 0; i-- {
            f[i] = f[i+1] - int64(skill[i+1])*int64(x)
        }
    }

    return f[n-1]
}

File: minimumTimeToBrewPotions.py
class Solution:
    def minTime(self, skill: List[int], mana: List[int]) -> int:
        n, m = len(skill), len(mana)
        dp = [0] * m

        for i in range(n):
            new_dp = [0] * m
            for j in range(m):
                time = skill[i] * mana[j]
                if i == 0 and j == 0:
                    new_dp[j] = time
                elif i == 0:
                    new_dp[j] = new_dp[j-1] + time
                elif j == 0:
                    new_dp[j] = dp[j] + time
                else:
                    new_dp[j] = max(dp[j], new_dp[j-1]) + time
            dp = new_dp

        return dp[-1]
